File: gocode/bitseal_web/bitseal.go
package bitseal

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"net/url"
	"sort"
	"strings"
	"time"

	"github.com/bsv-blockchain/go-sdk/message"
	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	crypto "github.com/bsv-blockchain/go-sdk/primitives/hash"
)

const ProtocolHeader = "BitSeal"

// RandomNonce returns 128-bit random hex string (32 characters)
func RandomNonce() (string, error) {
	buf := make([]byte, 16)
	_, err := rand.Read(buf)
	if err != nil {
		return "", err
	}
	return hex.EncodeToString(buf), nil
}

// CanonicalQueryString implements RFC3986 key-sort encoding
func CanonicalQueryString(q string) string {
	if q == "" {
		return ""
	}
	q = strings.TrimPrefix(q, "?")
	values, _ := url.ParseQuery(q)
	type kv struct{ k, v string }
	var kvs []kv
	for k := range values {
		for _, v := range values[k] {
			kvs = append(kvs, kv{k, v})
		}
	}
	sort.Slice(kvs, func(i, j int) bool { return kvs[i].k < kvs[j].k })
	var parts []string
	for _, pair := range kvs {
		parts = append(parts, url.QueryEscape(pair.k)+"="+url.QueryEscape(pair.v))
	}
	return strings.Join(parts, "&")
}

// BodyHashHex returns SHA256(body) in hex or empty string if body is empty
func BodyHashHex(body string) string {
	if body == "" {
		return ""
	}
	hash := crypto.Sha256([]byte(body))
	return hex.EncodeToString(hash)
}

// BuildCanonicalString joins components with newline
func BuildCanonicalString(method, uriPath, query, body, timestamp, nonce string) string {
	parts := []string{
		strings.ToUpper(method),
		uriPath,
		CanonicalQueryString(query),
		BodyHashHex(body),
		timestamp,
		nonce,
	}
	return strings.Join(parts, "\n")
}

// SignRequest constructs headers for a BitSeal request
func SignRequest(method, uriPath, query, body string, clientPriv *ec.PrivateKey, serverPub *ec.PublicKey) (map[string]string, error) {
	timestamp := time.Now().UnixMilli()
	nonce, err := RandomNonce()
	if err != nil {
		return nil, err
	}
	canonical := BuildCanonicalString(method, uriPath, query, body, fmt.Sprintf("%d", timestamp), nonce)
	digest := crypto.Sha256([]byte(canonical))
	sigBytes, err := message.Sign(digest, clientPriv, serverPub)
	if err != nil {
		return nil, err
	}
	sigBase64 := base64.StdEncoding.EncodeToString(sigBytes)
	headers := map[string]string{
		"X-BKSA-Protocol":  ProtocolHeader,
		"X-BKSA-Sig":       sigBase64,
		"X-BKSA-Timestamp": fmt.Sprintf("%d", timestamp),
		"X-BKSA-Nonce":     nonce,
	}
	return headers, nil
}

// VerifyRequest checks headers, returns true if signature OK
func VerifyRequest(method, uriPath, query, body string, headers map[string]string, serverPriv *ec.PrivateKey) (bool, error) {
	if headers["X-BKSA-Protocol"] != ProtocolHeader {
		return false, nil
	}
	timestamp := headers["X-BKSA-Timestamp"]
	nonce := headers["X-BKSA-Nonce"]
	sigBase64 := headers["X-BKSA-Sig"]
	if timestamp == "" || nonce == "" || sigBase64 == "" {
		return false, nil
	}
	canonical := BuildCanonicalString(method, uriPath, query, body, timestamp, nonce)
	// Verify signature against SHA256 digest (BRC-77)
	digest := crypto.Sha256([]byte(canonical))
	sigBytes, err := base64.StdEncoding.DecodeString(sigBase64)
	if err != nil {
		return false, err
	}
	ok, err := message.Verify(digest, sigBytes, serverPriv)
	if err != nil {
		return false, err
	}
	return ok, nil
}

File: tscode/bitseal_web/BitSeal.ts
// @ts-nocheck
// BitSeal protocol TypeScript helper

import { sha256 } from '@bsv/sdk/primitives/Hash'
import { toHex, toBase64, toArray } from '@bsv/sdk/primitives/utils'
import PrivateKey from '@bsv/sdk/primitives/PrivateKey'
import PublicKey from '@bsv/sdk/primitives/PublicKey'

import { sign as brc77Sign, verify as brc77Verify } from '@bsv/sdk/messages/SignedMessage'
import { randomBytes } from 'crypto'

export interface BitSealHeaders {
  'X-BKSA-Protocol': 'BitSeal'
  'X-BKSA-Sig': string
  'X-BKSA-Timestamp': string
  'X-BKSA-Nonce': string
  [k: string]: string
}

export const randomNonce = (): string => randomBytes(16).toString('hex')

export function canonicalQueryString (query: string = ''): string {
  if (!query) return ''
  if (query.startsWith('?')) query = query.slice(1)
  const params = new URLSearchParams(query)
  const tuples: Array<[string, string]> = []
  params.forEach((v, k) => tuples.push([k, v]))
  tuples.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0))
  return tuples.map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`).join('&')
}

export function bodyHashHex (body: string | undefined | null): string {
  if (!body) return ''
  return toHex(sha256(body, 'utf8'))
}

export function buildCanonicalString (
  method: string,
  uriPath: string,
  query: string,
  body: string,
  timestamp: string,
  nonce: string
): string {
  return [
    method.toUpperCase(),
    uriPath,
    canonicalQueryString(query),
    bodyHashHex(body),
    timestamp,
    nonce
  ].join('\n')
}

export function signRequest (
  method: string,
  uriPath: string,
  query: string,
  body: string,
  clientPriv: PrivateKey,
  serverPub: PublicKey,
  opts: Partial<{ timestamp: string, nonce: string }> = {}
): BitSealHeaders {
  const timestamp = opts.timestamp ?? Date.now().toString()
  const nonce = opts.nonce ?? randomNonce()
  const canonical = buildCanonicalString(method, uriPath, query, body, timestamp, nonce)
  const digest = sha256(canonical, 'utf8')
  const sigBytes = brc77Sign(digest, clientPriv, serverPub)
  const sigBase64 = toBase64(sigBytes)
  return {
    'X-BKSA-Protocol': 'BitSeal',
    'X-BKSA-Sig': sigBase64,
    'X-BKSA-Timestamp': timestamp,
    'X-BKSA-Nonce': nonce
  }
}

export function verifyRequest (
  method: string,
  uriPath: string,
  query: string,
  body: string,
  headers: Record<string, string>,
  serverPriv: PrivateKey
): boolean {
  // Header names may be lower-cased by fetch/bun runtime; normalize lookup helper
  const h = (name: string): string | undefined => {
    return headers[name] ?? headers[name.toLowerCase()] ?? headers[name.toUpperCase()]
  }

  if (h('X-BKSA-Protocol') !== 'BitSeal') return false
  const timestamp = h('X-BKSA-Timestamp')
  const nonce = h('X-BKSA-Nonce')
  const sigBase64 = h('X-BKSA-Sig')
  if (!timestamp || !nonce || !sigBase64) return false
  const canonical = buildCanonicalString(method, uriPath, query, body, timestamp, nonce)
  const digest = sha256(canonical, 'utf8')
  const sigBytes = toArray(sigBase64, 'base64')
  try {
    return brc77Verify(digest, sigBytes, serverPriv)
  } catch {
    return false
  }
} 